src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1084 @@
+                    ImmutableArray<BoundPositionalSubpattern> newSubpatterns = args.discards.SetItem(args.i, args.subpatternTemplate.WithPattern(newPattern));
+
+                    BoundPattern newRecursive = new BoundRecursivePattern(
+                        newPattern.Syntax, declaredType: args.recursiveTemplate.DeclaredType, deconstructMethod: args.recursiveTemplate.DeconstructMethod,
+                        deconstruction: newSubpatterns,
+                        properties: default, isExplicitNotNullTest: false, variable: null, variableAccess: null,
+                        args.recursiveTemplate.InputType, args.recursiveTemplate.NarrowedType, args.recursiveTemplate.HasErrors);
+
+                    newRecursive = WithInputTypeCheckIfNeeded(newRecursive, inputType);
+
+                    return newRecursive;
+                }
+
+                // Make `Type { Prop: <newPattern> }`
+                static BoundPattern makePropertyPattern(BoundPattern newPattern, TypeSymbol inputType,
+                    (BoundRecursivePattern recursiveTemplate, BoundPropertySubpattern subpatternTemplate) args)
+                {
+                    newPattern = WithInputTypeCheckIfNeeded(newPattern, args.subpatternTemplate.Pattern.InputType);
+                    ImmutableArray<BoundPropertySubpattern> newSubpatterns = [args.subpatternTemplate.WithPattern(newPattern)];
+
+                    BoundPattern newRecursive = new BoundRecursivePattern(
+                        newPattern.Syntax, declaredType: args.recursiveTemplate.DeclaredType, deconstructMethod: null, deconstruction: default,
+                        properties: newSubpatterns,
+                        isExplicitNotNullTest: false, variable: null, variableAccess: null,
+                        args.recursiveTemplate.InputType, args.recursiveTemplate.NarrowedType, args.recursiveTemplate.HasErrors);
+
+                    newRecursive = WithInputTypeCheckIfNeeded(newRecursive, inputType);
+
+                    return newRecursive;
+                }
+            }
+
+            private bool CanSkip(BoundPattern pattern)
+            {
+                if (_negated)
+                {
+                    return pattern is BoundNegatedPattern { Negated: BoundDiscardPattern };
+                }
+
+                return pattern is BoundDiscardPattern;
+            }
+
+            delegate BoundPattern ReplaceLeafPattern<T>(BoundPattern leaf, TypeSymbol inputType, T args);
+
+            // In a tree of `and` and `or` patterns, replace the leaves
+            //
+            // For example, when looking at the `(A or B) and C` in `{ Prop: (A or B) and C }`
+            // we produce `({ Prop: A } or { Prop: B }) and { Prop: C }`
+            // by using a replacement function (x) => `{ Prop: x }`.
+            private BoundPattern PullBinaryPatternsOut<TArgs>(BoundPattern pattern, TypeSymbol newInputType, ReplaceLeafPattern<TArgs> replaceLeaf, TArgs args)
```

Comment:
>PullBinaryPatternsOut [](http://example.com/codeflow?start=33&length=21)

I implemented a rough approach that eliminates a need for this helper and decreases the amount of performed input type adjustments. See the last commit in #76440.